projects/sandbox/train/tests/test_data_structures.py

```diff
@@ -0,0 +1,187 @@
+        kernel_size,
+        batch_size,
+        batches_per_epoch=batches_per_epoch,
+        preprocessor=None,
+        coincident=coincident,
+        shuffle=True,
+        device="cpu",
+    )
+
+    for i, (X, y) in enumerate(dataset):
+        assert X.shape == (batch_size, num_ifos, kernel_size)
+        assert y.shape == (batch_size, 1)
+        assert (y == 0).all().item()
+
+        for sample in X:
+            start = sample[0, 0]
+            expected = torch.arange(start, start + kernel_size)
+            assert (sample[0] == expected).all().item()
+            if num_ifos == 1:
+                continue
+
+            if coincident:
+                expected = expected + 128 * sample_rate
+                assert (sample[1] == expected).all().item()
+            else:
+                start = sample[1, 0]
+                expected = torch.arange(start, start + kernel_size)
+                assert (sample[1] == expected).all().item()
+    assert (i + 1) == batches_per_epoch
+
+
+def test_bbhnet_in_memory_dataloader_with_preprocessor(
+    sequential_data, sample_rate
+):
+    def preprocessor(X, y):
+        X[:, 0] *= 2
+        y[::2] = 1
+        return X, y
+
+    dataset = BBHInMemoryDataset(
+        sequential_data,
+        kernel_size=2 * sample_rate,
+        batch_size=8,
+        batches_per_epoch=2,
+        preprocessor=preprocessor,
+        coincident=True,
+        shuffle=True,
+        device="cpu",
+    )
+    for X, y in dataset:
+        assert ((X[:, 0] % 2) == 0).all().item()
+        if X.shape[1] > 1:
+            assert not ((X[:, 1] % 2) == 0).all().item()
+
+        assert (y[::2] == 1).all().item()
+        assert (y[1::2] == 0).all().item()
+
+
+@pytest.fixture(params=[0, 0.25, 1])
+def prob(request):
+    return request.param
+
+
+def get_rand_patch(size):
+    probs = torch.linspace(0, 0.99, size[1])
+    return torch.stack([probs] * size[0])
+
+
+def test_glitch_sampler(sample_rate, offset, device, prob):
+    glitches = torch.arange(512 * 2 * 10, dtype=torch.float32)
+    glitches = glitches.reshape(10, 512 * 2) + 1
+    sampler = GlitchSampler(
+        prob=prob,
+        max_offset=offset,
+        h1=glitches,
+        l1=-1 * glitches[:9],
+    )
+    sampler.to(device)
+    for glitch in sampler.buffers():
+        assert glitch.device.type == device
+
+    X = torch.zeros((8, 2, 512), dtype=torch.float32).to(device)
+    probs = get_rand_patch((2, 8)).to(device)
+    with patch("torch.rand", return_value=probs):
+        inserted, _ = sampler(X, None)
+
+        # TODO: the tests could be more extensive, but
+        # then are we functionally just testing sample_kernels?
+        if prob == 0:
+            assert (inserted == 0).all().item()
+        elif prob == 1:
+            assert (inserted != 0).all().item()
+        else:
+            assert (inserted[:2] != 0).all().item()
+            assert (inserted[2:] == 0).all().item()
+
+
+def sample(obj, N, dev=None):
```

Comment:
the `sample` method actually got rid of the `dev` kwarg so we can actually just drop this. FWIW `sampled_params` would have shape `(N, 3)` but since we don't actually use it for anything this doesn't really matter either.

Reply:
Sounds good, updated